src/rgw/driver/sfs/sqlite/dbconn.h

```diff
@@ -202,24 +204,64 @@ inline auto _make_storage(const std::string& path) {
 
 using Storage = decltype(_make_storage(""));
 
+static void sqlite_error_callback(void* ctx, int error_code, const char* msg) {
+  const auto cct = static_cast<CephContext*>(ctx);
+  lderr(cct) << "[SQLITE] (" << error_code << ") " << msg << dendl;
+}
+
+static int sqlite_profile_callback(
+    unsigned int reason, void* ctx, void* vstatement, void* runtime_ns
+) {
+  const auto cct = static_cast<CephContext*>(ctx);
+
+  if (reason == SQLITE_TRACE_PROFILE) {
+    const int64_t runtime = *static_cast<int64_t*>(runtime_ns) / 1000 / 1000;
+    sqlite3_stmt* statement = static_cast<sqlite3_stmt*>(vstatement);
+    const std::unique_ptr<char, void (*)(char*)> sql(
+        sqlite3_expanded_sql(statement),
+        [](char* p) { sqlite3_free(static_cast<void*>(p)); }
+    );
+    const sqlite3* db = sqlite3_db_handle(statement);
+
+    lderr(cct) << "[SQLITE PROFILE] " << std::hex << db << " " << runtime
+               << "ms "
+               << "\"" << sql.get() << "\"" << dendl;
+  }
+
+  return 0;
+}
+
 class DBConn {
   Storage storage;
 
  public:
   sqlite3* sqlite_db;
+  CephContext* const cct;
+  const bool profile_enabled;
 
-  DBConn(CephContext* cct) : storage(_make_storage(getDBPath(cct))) {
+  DBConn(CephContext* _cct)
+      : storage(_make_storage(getDBPath(_cct))),
```

Comment:
what if `_make_storage` fails?
To gain more control, could be better try to move `_make_storage` call in the constructor body? (or better, on an `init()` function).
I know that the pre-PR implementation already does this, but I was wondering if this could increase the overall safety.

Reply:
Would you please elaborate a bit what problem you would like to solve here?

Reply:
This was just a consideration about `_make_storage()` and what happens when (if) it fails.
But checking the code it is obvious that this function has designed to throw an exception since the beginning, so forget this comment :grin: